박스오피스 [STEP 3-1] Sunny, Blu

SunnnySong opened the pull request:
# 박스오피스 [STEP 3-1] Sunny, Blu
>안녕하세요! 븨븨(@YebinKim)!
🧴**TEAM SUNBLock**🧴 의 Sunny, Blu입니다!
>STEP3는 View를 만드는 3-1과 데이터와 View 연결하는 부분을 3-2로 나누어 해당 스텝을 두 단계로 나누어서 진행하려고 해요.
>
>우선, **STEP 3-1**에서는 **Mock 데이터를 구성하여 Collection View를 구현**해 보았어요.
>3-2에서는 네트워크 통신으로 받아온 데이터와 Collection View를 연결하는 세부 구현을 진행해 보려고 합니다!
>
>리뷰 잘 부탁드립니다! 감사합니다! ❤️

<br>

# 🧑🏼‍💻 작업 목록

## Mock 데이터를 활용한 Collection View 구현
- [x] Collection View 구현
    - [x] 리스트 형태로 박스오피스 정보 표기
    - [x] 맨 왼쪽: 영화의 현재 등수 표기
    - [x] 신규 영화: 등수 아래에 "신작" 표기
    - [x] 기존 영화: 어제와 비교한 등락을 표기
        - 순위 상승: 빨간 화살표 + 등락 편차
        - 순위 하락: 파란 화살표 + 등락 편차
        - 변동 없음: "-" 표기
- [x] Collection View에 보여질 모델 타입 구현
- [x] Safe Area을 고려한 오토 레이아웃 구현

<br>


# 📚 학습 키워드 및 학습 내용
## 1. Controller가 가벼운 MVC 구현하기

- **학습 내용**
   - 우리가 원래 알고 있었던 MVC는 Controller가 비대해진다는 단점을 갖고 있었음
   - 하지만 과거 븨븨에게 디자인 패턴에 대해 질문을 드렸을 때, Controller가 가벼운 MVC에 대한 힌트를 얻어 학습하게 되었음
- **적용 과정**
   - CollectionView의 DataSource 분리를 위한 BoxOfficeDataSource 타입 생성
       - 처음 구현했던 코드에는 UICollectionViewDataSource가 ViewController에 있었다. 
       - 하지만 DataSource는 UICollectionView라는 View에 데이터를 전달시키는 객체이지 ViewController에 있을 필요성을 느끼지 못했다.
       - BoxOfficeDataSource 타입을 따로 생성후 View 폴더에 넣고 ViewController를 더 가볍게 만들었다.
       
   - CollectionView의 layout을 생성하는 CollectionViewLayout 타입 생성
       - 처음 구현했던 코드에는 CollectionView layout을 만드는 함수가 View Controller 내부에 존재했었다.
       - CollectionView layout은 CollectionView라는 View를 만들기 위해 필요한 함수로, 위와 동일하게 View Controller에 있을 이유가 없다는 생각이 들었다.
       - CollectionViewLayout 타입을 생성해 View 폴더에 넣어 분리하였다.
  
- **고민**
   - 😰 **`UICollectionViewDataSource` vs `UICollectionViewDiffableDataSource` 선택**
        -  CollectionView가 iOS를 배우는데 있어어 중요한 학습 요소라 생각하였고 이번에 UICollectionView를 구현하면서 `UICollectionViewDataSource`와 `UICollectionViewDiffableDataSource` 모두 시도해 보려고 함
        - `UICollectionViewDiffableDataSource` 코드
    ```swift
        private func configureDataSource() {
        
        let cellRegistration = UICollectionView.CellRegistration<RankCell, RankItem> { (cell, indexPath, item) in
            
            cell.updateWithItem(item)
            cell.accessories = [.disclosureIndicator()]
        }
        
        dataSource = UICollectionViewDiffableDataSource<Section, RankItem>(collectionView: collectionView) { (collectionView: UICollectionView, indexPath: IndexPath, item: RankItem) -> UICollectionViewCell? in
            
            let collectionView = collectionView.dequeueConfiguredReusableCell(using: cellRegistration, for: indexPath, item: item)
            return collectionView
        }
        
        var snapshot = NSDiffableDataSourceSnapshot<Section, RankItem>()
        snapshot.appendSections([.main])
        snapshot.appendItems(RankItem.mock)
        
        dataSource?.apply(snapshot, animatingDifferences: false)
    }
    ```
    
    </br>
    
    - 🚀 **`UICollectionViewDataSource`로 결정한 이유**
        - UICollectionViewDiffableDataSource는 dataSource를 관리하는데 있어 더 깔끔한 코드를 제공해 줌
        - 데이터가 변경되었을 때 기존 `reloadData()`와 달리 애니메이션이 제공되 사용자 경험을 향상시킴
        - 하지만 우리의 앱은 네트워킹을 통해 데이터를 불러와 CollectionView에 전달해주면, 그 이후의 데이터 변경이 일어나지 않음
        - DiffableDataSource의 최대 장점은 손 쉬운 데이터 변경 및 업데이트 라고 생각하는데 우리의 프로젝트에서는 해당 장점을 활용할 수 없기 때문에 `UICollectionViewDataSource`를 사용했음

## 2. Cell에 데이터를 전달하는 방식
- **학습 내용**
    - 과거 CollectionViewCell에서 데이터를 연결할 때, DataSource의 `func collectionView(_, cellForItemAt)` 메서드에서 Cell의 요소와 데이터를 매칭해주는 코드를 구현했었음
    - UICollectionView를 학습함에 따라 `func collectionView(_, cellForItemAt)` 메서드는 CollectionView가 Cell을 표시할 때 마다 호출되기 때문에 내부 코드가 가벼워야 함을 알게 되었음

- **적용 과정**
    - UICollectionViewCell을 custom한 `CustomListCell` 에 프로퍼티로 cell에 보여질 데이터 타입을 갖도록 구현함.
    - `func updateConfiguration(using:)` 메서드 내부에서 cell의 UI Objects와 cell에 보여질 데이터 타입을 매칭시켰음.
    - `func collectionView(_, cellForItemAt)` 에서는 cell에 정의된 데이터 프로퍼티에 Mock 데이터의 요소들이 할당되도록 구현하였음.

## 3. 휴먼 에러를 줄이기 위한 Extension
- **고민**
    - CollectionView에서 Cell을 register 하거나 dequeue할 때, String 타입의 `Identifier`을 입력하는 과정에서 휴먼 에러가 발생할 가능성이 있다고 판단했음
- **적용 과정**
```swift
extension UICollectionReusableView {

    static var reuseIdentifier: String {
        return String(describing: Self.self)
    }
}    
```
```swift
extension UICollectionView {

    func register<T: UICollectionViewCell>(cell: T.Type) {

        register(T.self, forCellWithReuseIdentifier: T.reuseIdentifier)
    }

    func dequeue<T: UICollectionViewCell>(cell: T.Type, for indexPath: IndexPath) -> T {

        return dequeueReusableCell(withReuseIdentifier: cell.reuseIdentifier, for: indexPath) as! T
    }
}
```
- `UICollectionReusableView`에 extension으로 자기 자신의 타입을 String으로 갖는 `reuseIdentifier` 속성 추가
- `UICollectionView`에 extension으로 register와 dequeue 메서드 추가
    - 파라미터로 UICollectionViewCell의 메타타입을 입력하면, 해당 cell의 `reuseIdentifier`로 접근해 identifier을 넣어줌
    - identifier을 입력할 때 발생할 수 있는 휴먼 에러를 최대한 제거한 방법으로 구현했음

<br>

# 🔫 트러블 슈팅
## Separator layout guide

### 문제 발생
<img src="https://hackmd.io/_uploads/BkrQApdVh.png" width="400" height="400">

- (View 화면을 구현하고 있던 중이라,, 컬러플한 cell 이해 바랍니다,,,😅)
- Separator layout guide가 왼쪽 끝까지 붙지 않는 문제 발생

### 문제 해결 시도
- Apple 공식문서에 있는 Modern Collection Views 프로젝트에서 separatorLayoutGuide 키워드 힌트를 얻었다.
    - CustomCellListViewController.swift 내부의 updateSeparatorConstraint() 함수
    - <img src="https://hackmd.io/_uploads/rkLn0puV3.png" width="400" height="400">
    - 해당 프로젝트에서는 separatorLayoutGuide를 UIListContentView.textLayoutGuide.leadingAnchor 과 동일하게 주었다.
    ```swift
    private func updateSeparatorConstraint() {
        guard let textLayoutGuide = listContentView.textLayoutGuide else { return }
        if let existingConstraint = separatorConstraint, existingConstraint.isActive {
            return
        }
        let constraint = separatorLayoutGuide.leadingAnchor.constraint(equalTo: textLayoutGuide.leadingAnchor)
        constraint.isActive = true
        separatorConstraint = constraint
    }
    ```

### 문제 해결
```swift
let constraint = separatorLayoutGuide.leadingAnchor.constraint(equalTo: contentView.leadingAnchor)
constraint.isActive = true
```

calledBlu committed:
test: nil 값에 대한 워닝 발생 제거를 위한 테스트 코드 수정

- 문자열 보간법 사용 시 안에 존재하는 값이 nil값일 수 있는 가능성 때문에 에러 메세지 발생
- 이를 별도로 위에서 언래핑 후 언래핑된 값으로 문자열 보간법을 사용하도록 수정

SunnnySong committed:
refactor: NetworkProvider init에 default로 NetworkSession() 주입

SunnnySong committed:
Merge branch 'STEP2' of https://github.com/calledBlu/ios-box-office into STEP2

calledBlu committed:
test: test 수행 시 sut 값을 사용하도록 테스트 케이스 수정

- sut를 제거하는 경우 `setUpWithError()`, `tearDownWithError()`가 불필요해지는 문제 발생
- 일반적인 테스트 코드를 따라 테스트 케이스 내부에 sut를 사용하도록 sut에 apiKey 값을 할당하는 것으로 수정

SunnnySong committed:
feat: View에 보일 모델 타입 구현
- BoxOfficeItem 모델 타입 생성
- Rank 모델 타입 생성

calledBlu committed:
feat: 커스텀 셀 구현 및 타겟 최소버전 상향

- `View`그룹 생성
- 박스오피스 순위를 나타내는 `CustomListCell`을 `UICollectionViewListCell`을 채택하여 구현
- DTO를 기반으로 View에 데이터를 전달할 모델인 `BoxOfficeItem`을 내부 프로퍼티로 가지고 있도록 함
- `UILabel`, `UIStackView`를 활용한 셀 구성 및 Auto Layout 지정
- `configureSeparator()` 메서드로 하단 구분선이 왼쪽 뷰 끝까지 채워지도록 구현

SunnnySong committed:
feat: CollectionView 생성

- BoxOfficeViewController 에서 collectionView 생성
- BoxOfficeDataSource 생성
- CustomListCell에서 rankStackView constant 수정

calledBlu committed:
refactor: CollectionView, Cell, StackView 리팩토링

- `UIVerticalStackView` 타입을 구현하여 세로로 뷰가 연결되는 스택뷰 생성 코드를 단축하도록 리팩토링
- `UILabel`을 확장하여 `convenience init`을 통해 폰트 스타일을 지정하여 생성할 수 있도록 구현
- `UICollectionReusableView`를 확장하여 재사용할 cell의 reuseIdentifier를 직접 추론하도록 구현
- `UICollectionView`를 확장하여 cell의 타입으로 셀을 등록하고, reuseIdentifier를 가져오도록 하여 휴먼에러가 감소하도록 구현
- 위의 수정에 따른 `BoxOfficeDataSource`, `CustomListCell`, `BoxOfficeViewController` 코드 변경

SunnnySong committed:
refactor: 파일 폴더링 변경

- Network 폴더 -> Model 폴더로 이동
- CollectionViewLayout 파일 -> View 폴더로 이동
- BoxOfficeDataSource 파일 -> View 폴더로 이동

SunnnySong committed:
chore: Controller 폴더 생성

SunnnySong committed:
chore: CustomListCell 속성 이름 변경

calledBlu committed:
refactor: VC에서 collection view 분리 및 refreshControl 구현

- BoxOfficeViewController에서 collectionView를 생성하지 않고, 생성된 인스턴스를 가지고 있도록 `BoxOfficeCollectionView` 타입으로 분리
- BoxOfficeViewController를 확장하여 refresh control 구현

SunnnySong committed:
refactor: Constants 타입 생성

- CustomListCell의autoLayout을 잡을 때, constant를 위한 타입 생성 후 수정
- Rank에서 0 대신 .zero 수정

SunnnySong committed:
refactor: Constants 위치 변경

- 기존 Constants의 타입을 CustomListCell 내부로 변경

calledBlu committed:
refactor: 날짜의 타입을 구분하여 타입에 따른 format으로 변환하도록 수정

- Date Extension에서 날짜의 타입을 `network`, `title`로 나누는 `DateType`구현
- yesterday가 Date 타입의 어제 날짜를 가지고 있도록 수정
- `formatData(type:)`: 타입에 따라 다른 형식으로 변환하고 이를 String으로 변환하는 메서드

YebinKim commented on BoxOffice/Model/BoxOfficeRank/BoxOfficeItem.swift:
```diff
@@ -0,0 +1,22 @@
+    let name: String
+    let audience: String
+    
+    static let mock = [
+        BoxOfficeItem(rank: Rank(rank: "1", rankIntensity: "0", rankOldAndNew: .new), name: "경관의 피", audience: "오늘 64,050 / 총 69,228"),
+        BoxOfficeItem(rank: Rank(rank: "2", rankIntensity: "-3", rankOldAndNew: .old), name: "스파이더맨", audience: "오늘 64,050 / 총 69,228"),
+        BoxOfficeItem(rank: Rank(rank: "3", rankIntensity: "2", rankOldAndNew: .old), name: "씽2게더", audience: "오늘 64,050 / 총 69,228"),
+        BoxOfficeItem(rank: Rank(rank: "4", rankIntensity: "0", rankOldAndNew: .old), name: "씽2게더", audience: "오늘 64,050 / 총 69,228")
+    ]
```
질문;

샘플 데이터를 기존의 `box_office_sample.json` 데이터로 활용해볼 수도 있었을 거 같은데 하드코딩으로 데이터를 추가한 이유가 있을까요 🤔

SunnnySong replied:
엄.. `box_office_sample.json`을 추가해 Unit Test에만 사용하고 여기에 목 데이터로 사용할 생각은 해보지 못했네요...🥲 다음엔 목 데이터가 필요할 때 이부분에 대해 한번 더 생각해보도록 할게요!

YebinKim replied:
네엡! 대량의 데이터를 추가할 때 편하기도 해서 혹시 json 사용을 생각해보셨을지 궁금했어요 ㅎㅎ

YebinKim commented on BoxOffice/View/BoxOfficeRank/CustomListCell.swift:
```diff
@@ -0,0 +1,71 @@
+        rankDetailLabel.attributedText = boxOfficeItem?.rank.detail
+        movieNameLabel.text = boxOfficeItem?.name
+        movieDetailLabel.text = boxOfficeItem?.audience
+
```
p4;

불필요한 공백이 있어보여요 👀

SunnnySong replied:
수정했어요😀 븨븨의 꼼꼼함에 또 한번 감사드립니다😀

YebinKim commented on BoxOffice/Model/Network/Provider/Implementation/NetworkProvider.swift:
```diff
@@ -12,7 +12,7 @@ final class NetworkProvider: NetworkProvidable {
     private let deserializer = JSONDeserializer()
     
-    init(session: NetworkSessionProtocol) {
+    init(session: NetworkSessionProtocol = NetworkSession()) {
```
리뷰 코멘트 반영 고맙습니다! 😆

YebinKim reviewed (changes requested):
블루, 써니 안녕하세요 😊
Step3를 두 단계로 나눠서 구현하기로 하셨군요!
뷰와 데이터 연동 코드를 명확히 구분해보는 것은 좋은 시도라고 생각합니다 👍

이번 스텝은 UI 구현 위주라 코멘트가 많진 않은데요.
프로젝트 요구사항 스크린샷과 달리 현재는 하단 SafeArea 가 깨지고 있어서 이 부분만 수정하고 다음 스텝으로 진행해보면 좋을 거 같아요~
<img width="300" alt="SafeArea" src="https://github.com/tasty-code/ios-box-office/assets/38066039/e9425d52-f234-4bb0-8e28-00b8881def83">

그 외 기기별 사이즈 대응에는 문제 없는 것 확인했습니다!
고생 많으셨어요 ☺️

SunnnySong committed:
refactor: safeArea 반영한 CollectionView autoLayout 수정

- collectionView의 autoresizingMask 삭제
- collectionView에 safeArea 반영한 autoLayout 설정
- CustomListCell에 stackView leading, trailing constant 변경

SunnnySong commented:
### SafeArea 를 반영한 AutoLayout 수정

적절한 목 데이터를 사용하지 못해 SafeArea를 반영한 AutoLayout을 생각하지 못했어요. 븨븨의 코멘트를 받아 바로 수정했습니다!(븨븨의 대면 피드백과 함께,,,,💜💙)

- CollectionView에 적용했던 autoresizingMask 적용을 삭제했어요.
     - CollectionView의 위치와 크기를 결정하는데 있어 autoresizingMask를 사용하지 않고, autoLayout을 사용하는 것으로 변경했어요.
     - autoLayout을 사용함에 따라 `translatesAutoresizingMaskIntoConstraints = false`로 변경했어요.
- CollectionView의 leftAnchor, rightAnchor, bottomAnchor 를 `safeAreaLayoutGuide`를 사용해 잡았어요.

- **화면**
<img width="509" alt="스크린샷 2023-05-12 오후 4 50 49" src="https://github.com/tasty-code/ios-box-office/assets/85678496/4c2b0ed4-bb04-4c54-8017-e40f7ec05042">

calledBlu committed:
feat: 첫 화면 시작 시 로드 중임을 알 수 있도록 하는 Indicator 구현

- `UIActivityIndicatorView`를 채택한 `ActivityIndicatorView` 생성
- `UIVerticalStackView`와 `UIActivityIndicatorView`에 final 키워드 추가
- `BoxOfficeViewController`에 indicator 추가

SunnnySong committed:
feat: Presentation의 Dispatcher 타입 생성

- ViewDisplayable 타입 생성
- DTO의 데이터를 가져오고, PresentationModel로 변환하는 역할을 하는 PresentationDispatchable 타입 생
- PresentationDispatchable를 채택한 BoxOfficeDispatcher 생성

SunnnySong committed:
Merge branch 'STEP3' of https://github.com/calledBlu/ios-box-office into STEP3

SunnnySong committed:
Revert "Merge branch 'STEP3' of https://github.com/calledBlu/ios-box-office into STEP3"

This reverts commit 7383570212dfe50cc84676309e6971bd5ad000be, reversing
changes made to 95507d9df8d2d5fb846491a84e93ea2e6c79699e.

calledBlu committed:
feat: 첫 화면 시작 시 로드 중임을 알 수 있도록 하는 Indicator 구현

- `UIActivityIndicatorView`를 채택한 `ActivityIndicatorView` 생성
- `UIVerticalStackView`와 `UIActivityIndicatorView`에 final 키워드 추가
- `BoxOfficeViewController`에 indicator 추가

YebinKim reviewed (approved):
블루, 써니 안녕하세요~
Step3-1 수정까지 고생 많으셨어요!
이번 스텝은 여기서 머지 진행하도록 할게요 😄

다음 스텝 진행도 화이팅입니다!!
